# Direct RNA sequencing of mycobacteria

Peter Polgar, ISMB, University College London

Supervisors: Dr Kristine Arnvig, Dr Irilenia Nobeli

## Repository information

This repository contains the code used for the first rotation of my Lido PhD, involving direct RNA sequencing of *Mycobacterium smegmatis* with Oxford Nanopore's MinION platform and identification of rRNA processing intermediates. Short descriptions of the R code used to analyse the data and generate the figures are found below.

## Code

### rRNA_read_boundaries

Reads in the sam files generated by mapping the reads to the *Msm* genome, filters out reads mapping to *rrnA* and plots the start and end nucleotide of our reads compared to the start and end site of the mature processed 16S, 23S and 5S rRNA (Figure 8). Also shows these compared to the whole operon (Figure 7) and generates the length plots in Figure 4.

### rRNA_read_groups

Reads in the sam file, filters reads mapping to the *rrnA* operon and groups reads based on their start and end nucleotide location. Creates groups based on the co-occurrence of these two subgroups and filters out everything not present in decent numbers. Plots populous groups relative to the start end end site of mature rRNAs in the *rrnA* operon. Used for Figures 9 and 10.

### QC_graphs

Reads in the Guppy summary files generated from basecalling as tables, sorts reads between mycobacterial RNA and the spike-in enolase control based on a threshold of calibration strand identity of 70% (our Guppy curiously didn't generate a calibration score). Filters a out a sub-table with only the passed reads, generates data based on the summary file and plots them (Figure 5 in the report). Also plots cumulative yield for Figure 4.

### tidy_data

Creates a summary table of reads mapping to various gene features, amongst other things, for further steps.

### featurecounts_reads

Reads in the files generated by tidy_data.R, calculates statistics for them, and plot various QC plots based on reads mapping to various gene features (Figure 6)


## Acknowledgements
A huge thank you to my supervisors, Kristine and Irilenia, for helping me with the project, as well as the other people in the sRNA lab: Alex, Terry and Amy. You guys are ridiculously awesome!

Special thanks for Felix Gr√ºnberger, whose PhD project inspired this rotation, and whose code I *kindly appropriated* on many occasions for my project. Find his code [here](https://github.com/felixgrunberger/Native_RNAseq_Microbes) and his amazing paper on BioRxiv [here!](https://www.biorxiv.org/content/10.1101/2019.12.18.880849v2)

Finally a huge thank you for the Lido programme and the BBSRC for funding my personal and professional development!
